server/src/main/java/org/elasticsearch/index/cache/bitset/BitsetFilterCache.java

```diff
@@ -92,10 +96,20 @@ public BitsetFilterCache(IndexSettings indexSettings, Listener listener) {
             throw new IllegalArgumentException("listener must not be null");
         }
         this.index = indexSettings.getIndex();
-        this.loadRandomAccessFiltersEagerly = indexSettings.getValue(INDEX_LOAD_RANDOM_ACCESS_FILTERS_EAGERLY_SETTING);
+        this.loadRandomAccessFiltersEagerly = shouldLoadRandomAccessFiltersEagerly(indexSettings);
         this.listener = listener;
     }
 
+    static boolean shouldLoadRandomAccessFiltersEagerly(IndexSettings settings) {
+        boolean loadFiltersEagerlySetting = settings.getValue(INDEX_LOAD_RANDOM_ACCESS_FILTERS_EAGERLY_SETTING);
+        boolean hasIndexRole = DiscoveryNode.hasRole(settings.getNodeSettings(), DiscoveryNodeRole.INDEX_ROLE);
+        boolean isStateless = DiscoveryNode.isStateless(settings.getNodeSettings());
+        if (isStateless && hasIndexRole) {
```

Comment:
I added a check for stateless here.  Technically, the INDEX_FAST_REFRESH_SETTING can only be set to true in stateless so this check is redundant. However, I'd rather explicitly check this requirement here so that the safety of this is not dependent on validation logic elsewhere, should the code elsewhere ever change.